Annotation macros not expanded

Using the sbt-plugin, `mdoc` block with annotation macros fail to expand.
Macro paradise is present in the project dependencies.

Say you have a macro like:

```scala
  @compileTimeOnly("enable macro paradise to expand macro annotations")
  class SomeAnnotation(args: String*) extends StaticAnnotation {
    def macroTransform(annottees: Any*): Any = macro SomeAnnotation.impl
  }
```

Using in in a `mdoc` bloc will fail to compile and gives a `enable macro paradise to expand macro annotations` error.

Hey thanks for the reply!

The problem is that if I remove that, I get the following exception:

```
mdoc.internal.cli.FileException: /Users/julient/Documents/Spotify/scio/site/src/paradox/FAQ.md
Caused by: java.lang.ClassNotFoundException: repl.Session$
        at scala.reflect.internal.util.AbstractFileClassLoader.findClass(AbstractFileClassLoader.scala:72)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at mdoc.internal.markdown.MarkdownCompiler$.buildDocument(MarkdownCompiler.scala:46)
        at mdoc.internal.markdown.MdocPostProcessor.processScalaInputs(MdocPostProcessor.scala:136)
        at mdoc.internal.markdown.MdocPostProcessor.processDocument(MdocPostProcessor.scala:46)
        at com.vladsch.flexmark.parser.internal.PostProcessorManager.postProcess(PostProcessorManager.java:70)
        at com.vladsch.flexmark.parser.internal.PostProcessorManager.processDocument(PostProcessorManager.java:53)
        at com.vladsch.flexmark.parser.Parser.postProcess(Parser.java:389)
        at com.vladsch.flexmark.parser.Parser.parse(Parser.java:369)
        at mdoc.internal.markdown.Markdown$.toDocument(Markdown.scala:81)
        at mdoc.internal.markdown.Markdown$.toMarkdown(Markdown.scala:113)
        at mdoc.internal.cli.MainOps.handleMarkdown(MainOps.scala:64)
        at mdoc.internal.cli.MainOps.handleFile(MainOps.scala:91)
        at mdoc.internal.cli.MainOps.$anonfun$generateCompleteSite$1(MainOps.scala:130)
        at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126)
        at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122)
        at scala.collection.immutable.List.foldLeft(List.scala:89)
        at mdoc.internal.cli.MainOps.generateCompleteSite(MainOps.scala:128)
        at mdoc.internal.cli.MainOps.run(MainOps.scala:149)
        at mdoc.internal.cli.MainOps$.process(MainOps.scala:231)
        at mdoc.Main$.process(Main.scala:26)
        at mdoc.Main$.process(Main.scala:21)
        at mdoc.Main$.main(Main.scala:16)
        at mdoc.Main.main(Main.scala)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at sbt.Run.invokeMain(Run.scala:98)
        at sbt.Run.run0(Run.scala:92)
        at sbt.Run.execute$1(Run.scala:68)
        at sbt.Run.$anonfun$run$4(Run.scala:80)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
        at sbt.util.InterfaceUtil$$anon$1.get(InterfaceUtil.scala:10)
        at sbt.TrapExit$App.run(TrapExit.scala:253)
        at java.lang.Thread.run(Thread.java:748)
mdoc.internal.cli.FileException: /Users/julient/Documents/Spotify/scio/site/src/paradox/internals/Coders.md
Caused by: java.lang.ClassNotFoundException: repl.Session$
        at scala.reflect.internal.util.AbstractFileClassLoader.findClass(AbstractFileClassLoader.scala:72)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at mdoc.internal.markdown.MarkdownCompiler$.buildDocument(MarkdownCompiler.scala:46)
        at mdoc.internal.markdown.MdocPostProcessor.processScalaInputs(MdocPostProcessor.scala:136)
        at mdoc.internal.markdown.MdocPostProcessor.processDocument(MdocPostProcessor.scala:46)
        at com.vladsch.flexmark.parser.internal.PostProcessorManager.postProcess(PostProcessorManager.java:70)
        at com.vladsch.flexmark.parser.internal.PostProcessorManager.processDocument(PostProcessorManager.java:53)
        at com.vladsch.flexmark.parser.Parser.postProcess(Parser.java:389)
        at com.vladsch.flexmark.parser.Parser.parse(Parser.java:369)
        at mdoc.internal.markdown.Markdown$.toDocument(Markdown.scala:81)
        at mdoc.internal.markdown.Markdown$.toMarkdown(Markdown.scala:113)
        at mdoc.internal.cli.MainOps.handleMarkdown(MainOps.scala:64)
        at mdoc.internal.cli.MainOps.handleFile(MainOps.scala:91)
        at mdoc.internal.cli.MainOps.$anonfun$generateCompleteSite$1(MainOps.scala:130)
        at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126)
        at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122)
        at scala.collection.immutable.List.foldLeft(List.scala:89)
        at mdoc.internal.cli.MainOps.generateCompleteSite(MainOps.scala:128)
        at mdoc.internal.cli.MainOps.run(MainOps.scala:149)
        at mdoc.internal.cli.MainOps$.process(MainOps.scala:231)
        at mdoc.Main$.process(Main.scala:26)
        at mdoc.Main$.process(Main.scala:21)
        at mdoc.Main$.main(Main.scala:16)
        at mdoc.Main.main(Main.scala)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at sbt.Run.invokeMain(Run.scala:98)
        at sbt.Run.run0(Run.scala:92)
        at sbt.Run.execute$1(Run.scala:68)
        at sbt.Run.$anonfun$run$4(Run.scala:80)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
        at sbt.util.InterfaceUtil$$anon$1.get(InterfaceUtil.scala:10)
        at sbt.TrapExit$App.run(TrapExit.scala:253)
        at java.lang.Thread.run(Thread.java:748)
info: Compiled in 20.66s (0 errors)
```

I've tried setting `scalacOptions := Seq("-Ydelambdafy:inline")` in the project, but it does not seem to work. I think `scalacOptions` is ignored.

--- a/mdoc/src/main/scala/mdoc/internal/cli/Settings.scala
+++ b/mdoc/src/main/scala/mdoc/internal/cli/Settings.scala
@@ -73,7 +73,8 @@ case class Settings(
       "Compiler flags such as compiler plugins '-Xplugin:kind-projector.jar' " +
         "or custom options '-deprecated'. Formatted as a single string with space separated values. " +
         "To pass multiple values: --scalac-options \"-Yrangepos -deprecated\". " +
-        "Defaults to the value of 'scalacOptions' in the 'mdoc.properties' resource file, if any."
+        "Defaults to the value of 'scalacOptions' in the 'mdoc.properties' resource file, if any. " +
+        "When using sbt-mdoc, update the `scalacOptions` sbt setting instead of passing --scalac-options to `mdocExtraArguments`."
     )
     scalacOptions: String = "",
     @Description("Remove all files in the output directory before generating a new site.")

--- a/mdoc/src/main/scala/mdoc/internal/markdown/MarkdownCompiler.scala
+++ b/mdoc/src/main/scala/mdoc/internal/markdown/MarkdownCompiler.scala
@@ -41,9 +41,8 @@ object MarkdownCompiler {
     val instrumentedInput = InstrumentedInput(filename, instrumented)
     val compileInput = Input.VirtualFile(filename, instrumented)
     val edit = TokenEditDistance.fromTrees(sectionInputs.map(_.source), compileInput)
-    val doc = compiler.compile(compileInput, reporter, edit) match {
-      case Some(loader) =>
-        val cls = loader.loadClass("repl.Session$")
+    val doc = compiler.compile(compileInput, reporter, edit, "repl.Session$") match {
+      case Some(cls) =>
         val ctor = cls.getDeclaredConstructor()
         ctor.setAccessible(true)
         val doc = ctor.newInstance().asInstanceOf[DocumentBuilder].$doc
@@ -119,7 +118,10 @@ class MarkdownCompiler(
   settings.processArgumentString(scalacOptions)
 
   private val sreporter = new FilterStoreReporter(settings)
-  val global = new Global(settings, sreporter)
+  var global = new Global(settings, sreporter)
+  private def reset(): Unit = {
+    global = new Global(settings, sreporter)
+  }
   private val appClasspath: Array[URL] = classpath
     .split(File.pathSeparator)
     .map(path => new File(path).toURI.toURL)
@@ -140,7 +142,8 @@ class MarkdownCompiler(
 
   def fail(original: Seq[Tree], input: Input, sectionPos: Position): String = {
     sreporter.reset()
-    val run = new global.Run
+    val g = global
+    val run = new g.Run
     run.compileSources(List(toSource(input)))
     val out = new ByteArrayOutputStream()
     val ps = new PrintStream(out)
@@ -161,15 +164,38 @@ class MarkdownCompiler(
   def compileSources(input: Input, vreporter: Reporter, edit: TokenEditDistance): Unit = {
     clearTarget()
     sreporter.reset()
-    val run = new global.Run
+    val g = global
+    val run = new g.Run
     run.compileSources(List(toSource(input)))
     report(vreporter, input, edit)
   }
 
-  def compile(input: Input, vreporter: Reporter, edit: TokenEditDistance): Option[ClassLoader] = {
+  def compile(
+      input: Input,
+      vreporter: Reporter,
+      edit: TokenEditDistance,
+      className: String,
+      retry: Int = 0
+  ): Option[Class[_]] = {
     compileSources(input, vreporter, edit)
     if (!sreporter.hasErrors) {
-      Some(new AbstractFileClassLoader(target, appClassLoader))
+      val loader = new AbstractFileClassLoader(target, appClassLoader)
+      try {
+        Some(loader.loadClass(className))
+      } catch {
+        case _: ClassNotFoundException =>
+          if (retry < 1) {
+            reset()
+            compile(input, vreporter, edit, className, retry + 1)
+          } else {
+            vreporter.error(
+              s"${input.syntax}: skipping file, the compiler produced no classfiles " +
+                "and reported no errors to explain what went wrong during compilation. " +
+                "Please report an issue to https://github.com/scalameta/mdoc/issues."
+            )
+            None
+          }
+      }
     } else {
       None
     }

